Spike: Extracted Payment Process

RadoslavGeorgiev opened the pull request:
This is an exploration of WCPay's payment process, meant to combine similar steps from different payment processes.

RadoslavGeorgiev committed:
Initial implementation.

Right now there is an order payment class, but I do not like how it is loaded. Ideally there should be an order payment class, but it should not manage the link between itself and the order. In the next iteration, that will be separate to support a scenario where an order could potentially have multiple payments.

RadoslavGeorgiev committed:
Adding the flexibility to have multiple payments stored for an order.

RadoslavGeorgiev committed:
Removing the need for a meta key, using directories instead

RadoslavGeorgiev committed:
Adding payment method objects to the mix.

RadoslavGeorgiev committed:
Adding flags

RadoslavGeorgiev committed:
Adding flags

RadoslavGeorgiev committed:
Adding payment steps, incl. free orders

RadoslavGeorgiev committed:
Adding payment steps, incl. free orders

RadoslavGeorgiev committed:
Adding payment steps, incl. free orders

RadoslavGeorgiev committed:
First successful order (even without payment)

RadoslavGeorgiev committed:
Prettifying code

RadoslavGeorgiev committed:
Temp work for a simple payment

RadoslavGeorgiev committed:
Merging develop

RadoslavGeorgiev committed:
Adding a step for saving the payment method, minor changes in the standard payment step.

RadoslavGeorgiev committed:
Completing the rest of the standard payment step, addin the update order step.

RadoslavGeorgiev committed:
Add a step to schedule payment method updates, and improve the store metadata step.

RadoslavGeorgiev committed:
Twaking the no-payment step.

botwoo commented:
### Test the build on Jurassic Ninja 

This feature is only available for A12s. Please log in into https://jurassic.ninja/specialops/ before clicking the link below:

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=spike/payment-process-structure) :rocket:

- Latest commit: f9f6644ae9bf5abcf77af2c21fb3fb0f02d091c6
- Build time: 2023-03-22 18:31:20 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev committed:
Adding split-UPE payment methods to the standard payment step.

RadoslavGeorgiev committed:
Adding the setup payment step.

RadoslavGeorgiev committed:
Working simple payments, incl. saving PMs

RadoslavGeorgiev committed:
Adding a bunch of UPE basics.

Fatal errors over the place, but there are too many uncommited changes.

RadoslavGeorgiev committed:
Slowly chugging through the UPE process_redirect_payment

RadoslavGeorgiev committed:
Finished with the redirect UPE step class

RadoslavGeorgiev committed:
Adding payment method type getters to intents

RadoslavGeorgiev committed:
Adding the last missing steps

RadoslavGeorgiev committed:
Verifying that steps do not have syntax errors

RadoslavGeorgiev committed:
UPE intent creation working, even though it is not being stored in session ATM.

RadoslavGeorgiev committed:
Working UPE process_payment

RadoslavGeorgiev committed:
Working UPE simple payment

RadoslavGeorgiev committed:
Fixing the standard checkout

RadoslavGeorgiev committed:
Adding flow to the payment, to clean up things

RadoslavGeorgiev committed:
Working post-checkout redirect

RadoslavGeorgiev committed:
Making sure that flows must be set

RadoslavGeorgiev committed:
Strong-typing vars

RadoslavGeorgiev committed:
Introducing a logger, and moving setters/getters into a separate trait

RadoslavGeorgiev committed:
Loading the existing payment process in the standard gateway, logging changes.

RadoslavGeorgiev committed:
Converting the logger to a normal class.

RadoslavGeorgiev committed:
Switching to existing payments, instead of creating new ones

RadoslavGeorgiev committed:
Isolating step execution into a separate method

RadoslavGeorgiev committed:
Adding payment status to the mix

RadoslavGeorgiev committed:
First use of the payment status

RadoslavGeorgiev committed:
Subs renewals now use the new payment process as well

RadoslavGeorgiev committed:
Removing step IDs and making steps final

RadoslavGeorgiev committed:
Unifying logic into is_applicable()

RadoslavGeorgiev committed:
Working WooPay checkout

Checkign out through WooPay works. Saving PMs there is a bit weird, as server-side everything looks well, but the PM does not appear on WooPay.

RadoslavGeorgiev commented:
A better version of the payment process prototype can be found in https://github.com/Automattic/woocommerce-payments/pull/6287. Closing this PR...
